File: src/ios/UI/Icons.swift
import SwiftUI

// Append-only v1 IDs and bundled vector geometry match Android Icons.kt.
let meshAnimals = ["mask", "cat", "dog", "lizard", "raccoon", "turtle", "fish", "bird", "whale"]
let meshReactions = ["fire", "heart", "laugh", "thumb", "party", "surprise", "melt", "speaker"]
let meshPalette: [UInt32] = [0x5DCAA5,0xF49B83,0xBDA6EF,0xE8A0CE,0x8DBBEF,0xA3CB8C,0xEFBB67,0x91CCCA,0x9FE1CB,0xF7BEAA,0xD3C5F5,0xF1C1DF,0xB9D4F6,0xC5DEB3,0xF5D79E,0xBCE2E0]

struct MeshIcon: View {
    let name: String
    var body: some View { MeshSymbol(name: name).stroke(style: StrokeStyle(lineWidth: 1.7, lineCap: .round, lineJoin: .round)).frame(width: 28, height: 28).accessibilityLabel(name) }
}
struct MeshAvatar: View {
    let value: UInt8
    @Environment(\.colorScheme) private var scheme
    private var tint: Color {
        let rgb = value == 0 ? 0xB4B2A9 : meshPalette[Int(value >> 4)]
        let scale = scheme == .light ? 0.45 : 1.0
        return Color(red: Double((rgb >> 16) & 255) / 255 * scale, green: Double((rgb >> 8) & 255) / 255 * scale, blue: Double(rgb & 255) / 255 * scale)
    }
    var body: some View { MeshIcon(name: Int(value & 15) < meshAnimals.count ? meshAnimals[Int(value & 15)] : "paw").foregroundColor(tint).padding(8).background(tint.opacity(0.12)).clipShape(Circle()) }
}
private struct MeshSymbol: Shape {
    let name: String
    func path(in rect: CGRect) -> Path {
        var p = Path()
        func m(_ x: CGFloat, _ y: CGFloat) { p.move(to: CGPoint(x: x, y: y)) }
        func l(_ x: CGFloat, _ y: CGFloat) { p.addLine(to: CGPoint(x: x, y: y)) }
        func c(_ a: CGFloat,_ b: CGFloat,_ d: CGFloat,_ e: CGFloat,_ x: CGFloat,_ y: CGFloat) { p.addCurve(to: CGPoint(x: x,y: y),control1: CGPoint(x: a,y: b),control2: CGPoint(x: d,y: e)) }
        func q(_ a: CGFloat,_ b: CGFloat,_ x: CGFloat,_ y: CGFloat) { p.addQuadCurve(to: CGPoint(x: x,y: y),control: CGPoint(x: a,y: b)) }
        func o(_ x: CGFloat,_ y: CGFloat,_ r: CGFloat) { p.addEllipse(in: CGRect(x: x-r,y: y-r,width: r*2,height: r*2)) }
        switch name {
        case "wave": m(2,13);c(6,3,9,22,13,12);c(17,2,19,17,22,9)
        case "lightning-bolt": m(14,2);l(4,14);l(11,14);l(10,22);l(20,9);l(13,9);p.closeSubpath()
        case "fire": m(12,2);c(17,8,23,13,18,19);c(9,27,1,16,8,9);c(7,14,15,10,12,2);p.closeSubpath()
        case "heart": m(12,21);c(-7,8,7,-1,12,7);c(18,-1,31,8,12,21);p.closeSubpath()
        case "moon": m(17,3);c(1,-1,1,26,21,18);c(9,20,7,7,17,3);p.closeSubpath()
        case "star": m(12,2);l(16,9);l(22,12);l(16,15);l(12,22);l(8,15);l(2,12);l(8,9);p.closeSubpath()
        case "crystal": m(12,2);l(20,8);l(17,20);l(7,22);l(4,9);p.closeSubpath();m(12,2);l(10,10);l(7,22);m(10,10);l(20,8)
        case "sun": o(12,12,5);for i in 0...7 { let a = CGFloat(i) * .pi / 4; m(12+8*cos(a),12+8*sin(a));l(12+11*cos(a),12+11*sin(a)) }
        case "spiral": m(12,12);for i in 1...120 { let a=CGFloat(i)*0.12,r=CGFloat(i)*0.075;l(12+r*cos(a),12+r*sin(a)) }
        case "disco-ball": o(12,13,9);m(12,0);l(12,4);for x in [CGFloat(8),12,16] {m(x,6);l(x,20)};for y in [CGFloat(9),13,17] {m(5,y);l(19,y)}
        case "open-lock", "closed-lock": m(7,11);l(7,7);c(7,0,18,0,18,7);if name == "closed-lock" {l(18,11)};m(4,11);l(21,11);l(21,22);l(4,22);p.closeSubpath()
        case "mushroom": m(2,14);c(3,-1,22,-1,23,14);p.closeSubpath();m(10,14);l(9,22);l(16,22);l(15,14)
        case "cactus": m(10,22);l(10,4);c(10,0,15,0,15,4);l(15,22);m(10,13);l(5,13);l(5,7);m(15,16);l(20,16);l(20,10)
        case "speaker": m(3,9);l(8,9);l(14,4);l(14,21);l(8,16);l(3,16);p.closeSubpath();m(18,6);c(25,9,25,16,18,19)
        case "mask": m(2,7);q(12,3,22,7);q(21,21,12,15);q(3,21,2,7);p.closeSubpath();o(7,11,1);o(17,11,1)
        case "cat", "dog", "raccoon":
            o(12,13,8)
            if name == "cat" || name == "raccoon" {m(5,9);l(3,2);l(10,6);m(14,6);l(21,2);l(19,9)}
            if name == "dog" {m(5,6);q(-3,14,5,17);m(19,6);q(28,14,19,17)}
            if name == "raccoon" {m(5,10);l(19,10);l(16,15);l(8,15);p.closeSubpath()}
            o(9,12,1);o(15,12,1);m(9,17);q(12,19,15,17)
        case "lizard": m(12,3);c(3,6,18,14,10,20);q(6,24,4,19);m(10,8);l(4,8);l(2,5);m(12,11);l(20,8);l(22,5);m(11,15);l(18,17);l(19,21);o(12,4,2)
        case "turtle": p.addEllipse(in: CGRect(x:5,y:7,width:14,height:13));o(12,4,3);m(6,9);l(2,6);m(18,9);l(22,6);m(7,18);l(3,22);m(17,18);l(21,22);m(12,7);l(8,13);l(12,19);l(16,13);p.closeSubpath()
        case "fish": m(2,12);q(9,1,17,9);l(23,5);l(23,19);l(17,15);q(9,23,2,12);p.closeSubpath();o(7,11,1)
        case "bird": m(4,20);l(4,10);c(4,1,17,0,18,9);l(23,12);l(18,14);q(16,23,4,20);m(6,14);q(11,21,15,13);o(14,8,1)
        case "whale": m(2,12);c(2,1,17,5,17,14);l(22,9);l(22,17);c(10,25,2,21,2,12);p.closeSubpath();m(8,5);l(8,1);m(8,3);l(12,1);o(6,12,1)
        case "paw": p.addEllipse(in: CGRect(x:6,y:12,width:12,height:9));for i in 0...3 {o(CGFloat(4+i*5),i == 0 || i == 3 ? 9 : 5,2)}
        case "thumb": m(5,10);l(10,10);l(13,2);l(16,3);l(15,10);l(22,10);l(20,21);l(5,21);p.closeSubpath()
        case "party": m(2,22);l(7,7);l(17,17);p.closeSubpath();m(10,4);l(13,1);m(17,9);l(22,5);m(20,14);l(24,13)
        case "comet": o(7,17,5);m(3,12);l(21,2);l(12,21);m(11,13);l(21,3)
        case "surprise": o(12,12,9);o(8,9,1);o(16,9,1);o(12,16,3)
        case "laugh": o(12,12,9);m(6,10);l(8,8);l(10,10);m(14,10);l(16,8);l(18,10);m(6,14);l(18,14);q(12,25,6,14)
        case "melt": m(3,15);c(-1,0,24,0,21,15);c(28,19,19,24,13,20);c(5,26,-3,20,3,15);o(8,9,1);o(16,11,1);m(8,15);q(13,20,17,14)
        default: o(12,12,8);o(9,10,1);o(15,10,1);m(8,15);q(12,20,16,15)
        }
        return p.applying(CGAffineTransform(scaleX: rect.width/24, y: rect.height/24).translatedBy(x: rect.minX, y: rect.minY))
    }
}

File: src/android/ui/src/main/kotlin/org/meshchat/ui/Icons.kt
package org.meshchat.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.luminance
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.scale
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp

internal val animals = listOf("mask","cat","dog","lizard","raccoon","turtle","fish","bird","whale")
internal val palette = listOf(0xFF5DCAA5,0xFFF49B83,0xFFBDA6EF,0xFFE8A0CE,0xFF8DBBEF,0xFFA3CB8C,0xFFEFBB67,0xFF91CCCA,
    0xFF9FE1CB,0xFFF7BEAA,0xFFD3C5F5,0xFFF1C1DF,0xFFB9D4F6,0xFFC5DEB3,0xFFF5D79E,0xFFBCE2E0).map { Color(it) }
internal val reactions = listOf("fire","heart","laugh","thumb","party","surprise","melt","speaker")
@Composable
internal fun nicknameColor(sender: ByteArray): Color {
    val base=palette[sender.fold(0){sum,b -> (sum*31+(b.toInt() and 255)) and 15}]
    return if(MaterialTheme.colorScheme.background.luminance()>.5f) Color(base.red*.45f,base.green*.45f,base.blue*.45f) else base
}
/** Bundled vector paths; never platform emoji or attacker-provided images. */
@Composable
fun MeshIcon(name: String, tint: Color = MaterialTheme.colorScheme.primary, modifier: Modifier = Modifier) {
    Canvas(modifier.size(28.dp).semantics { contentDescription = name }) {
        val stroke=Stroke(width=1.7f)
        scale(size.width/24f,size.height/24f,pivot=Offset.Zero) {
            val p=Path()
            when(name) {
                "wave" -> { p.moveTo(2f,13f);p.cubicTo(6f,3f,9f,22f,13f,12f);p.cubicTo(17f,2f,19f,17f,22f,9f) }
                "lightning-bolt" -> { p.moveTo(14f,2f);p.lineTo(4f,14f);p.lineTo(11f,14f);p.lineTo(10f,22f);p.lineTo(20f,9f);p.lineTo(13f,9f);p.close() }
                "fire" -> { p.moveTo(12f,2f);p.cubicTo(17f,8f,23f,13f,18f,19f);p.cubicTo(9f,27f,1f,16f,8f,9f);p.cubicTo(7f,14f,15f,10f,12f,2f);p.close() }
                "heart" -> { p.moveTo(12f,21f);p.cubicTo(-7f,8f,7f,-1f,12f,7f);p.cubicTo(18f,-1f,31f,8f,12f,21f);p.close() }
                "moon" -> { p.moveTo(17f,3f);p.cubicTo(1f,-1f,1f,26f,21f,18f);p.cubicTo(9f,20f,7f,7f,17f,3f);p.close() }
                "star" -> { p.moveTo(12f,2f);p.lineTo(16f,9f);p.lineTo(22f,12f);p.lineTo(16f,15f);p.lineTo(12f,22f);p.lineTo(8f,15f);p.lineTo(2f,12f);p.lineTo(8f,9f);p.close() }
                "crystal" -> { p.moveTo(12f,2f);p.lineTo(20f,8f);p.lineTo(17f,20f);p.lineTo(7f,22f);p.lineTo(4f,9f);p.close();p.moveTo(12f,2f);p.lineTo(10f,10f);p.lineTo(7f,22f);p.moveTo(10f,10f);p.lineTo(20f,8f) }
                "sun" -> { drawCircle(tint,5f,Offset(12f,12f),style=stroke);for(i in 0..7){val a=i*Math.PI/4;p.moveTo(12f+8f*kotlin.math.cos(a).toFloat(),12f+8f*kotlin.math.sin(a).toFloat());p.lineTo(12f+11f*kotlin.math.cos(a).toFloat(),12f+11f*kotlin.math.sin(a).toFloat())} }
                "spiral" -> { p.moveTo(12f,12f);for(i in 1..120){val a=i*.12f;val r=i*.075f;p.lineTo(12f+r*kotlin.math.cos(a),12f+r*kotlin.math.sin(a))} }
                "disco-ball" -> { drawCircle(tint,9f,Offset(12f,13f),style=stroke);p.moveTo(12f,0f);p.lineTo(12f,4f);for(x in listOf(8f,12f,16f)){p.moveTo(x,6f);p.lineTo(x,20f)};for(y in listOf(9f,13f,17f)){p.moveTo(5f,y);p.lineTo(19f,y)} }
                "open-lock","closed-lock" -> { p.moveTo(7f,11f);p.lineTo(7f,7f);p.cubicTo(7f,0f,18f,0f,18f,7f);if(name=="closed-lock")p.lineTo(18f,11f);p.moveTo(4f,11f);p.lineTo(21f,11f);p.lineTo(21f,22f);p.lineTo(4f,22f);p.close() }
                "mushroom" -> { p.moveTo(2f,14f);p.cubicTo(3f,-1f,22f,-1f,23f,14f);p.close();p.moveTo(10f,14f);p.lineTo(9f,22f);p.lineTo(16f,22f);p.lineTo(15f,14f) }
                "cactus" -> { p.moveTo(10f,22f);p.lineTo(10f,4f);p.cubicTo(10f,0f,15f,0f,15f,4f);p.lineTo(15f,22f);p.moveTo(10f,13f);p.lineTo(5f,13f);p.lineTo(5f,7f);p.moveTo(15f,16f);p.lineTo(20f,16f);p.lineTo(20f,10f) }
                "speaker" -> { p.moveTo(3f,9f);p.lineTo(8f,9f);p.lineTo(14f,4f);p.lineTo(14f,21f);p.lineTo(8f,16f);p.lineTo(3f,16f);p.close();p.moveTo(18f,6f);p.cubicTo(25f,9f,25f,16f,18f,19f) }
                "mask" -> { p.moveTo(2f,7f);p.quadraticTo(12f,3f,22f,7f);p.quadraticTo(21f,21f,12f,15f);p.quadraticTo(3f,21f,2f,7f);p.close();drawCircle(tint,1f,Offset(7f,11f));drawCircle(tint,1f,Offset(17f,11f)) }
                "cat","dog","raccoon" -> {
                    drawCircle(tint,8f,Offset(12f,13f),style=stroke)
                    if(name=="cat"||name=="raccoon"){p.moveTo(5f,9f);p.lineTo(3f,2f);p.lineTo(10f,6f);p.moveTo(14f,6f);p.lineTo(21f,2f);p.lineTo(19f,9f)}
                    if(name=="dog"){p.moveTo(5f,6f);p.quadraticTo(-3f,14f,5f,17f);p.moveTo(19f,6f);p.quadraticTo(28f,14f,19f,17f)}
                    if(name=="raccoon"){p.moveTo(5f,10f);p.lineTo(19f,10f);p.lineTo(16f,15f);p.lineTo(8f,15f);p.close()}
                    drawCircle(tint,1f,Offset(9f,12f));drawCircle(tint,1f,Offset(15f,12f));p.moveTo(9f,17f);p.quadraticTo(12f,19f,15f,17f)
                }
                "lizard" -> { p.moveTo(12f,3f);p.cubicTo(3f,6f,18f,14f,10f,20f);p.quadraticTo(6f,24f,4f,19f);p.moveTo(10f,8f);p.lineTo(4f,8f);p.lineTo(2f,5f);p.moveTo(12f,11f);p.lineTo(20f,8f);p.lineTo(22f,5f);p.moveTo(11f,15f);p.lineTo(18f,17f);p.lineTo(19f,21f);drawCircle(tint,2f,Offset(12f,4f)) }
                "turtle" -> { drawOval(tint,Offset(5f,7f),androidx.compose.ui.geometry.Size(14f,13f),style=stroke);drawCircle(tint,3f,Offset(12f,4f),style=stroke);p.moveTo(6f,9f);p.lineTo(2f,6f);p.moveTo(18f,9f);p.lineTo(22f,6f);p.moveTo(7f,18f);p.lineTo(3f,22f);p.moveTo(17f,18f);p.lineTo(21f,22f);p.moveTo(12f,7f);p.lineTo(8f,13f);p.lineTo(12f,19f);p.lineTo(16f,13f);p.close() }
                "fish" -> { p.moveTo(2f,12f);p.quadraticTo(9f,1f,17f,9f);p.lineTo(23f,5f);p.lineTo(23f,19f);p.lineTo(17f,15f);p.quadraticTo(9f,23f,2f,12f);p.close();drawCircle(tint,1f,Offset(7f,11f)) }
                "bird" -> { p.moveTo(4f,20f);p.lineTo(4f,10f);p.cubicTo(4f,1f,17f,0f,18f,9f);p.lineTo(23f,12f);p.lineTo(18f,14f);p.quadraticTo(16f,23f,4f,20f);p.moveTo(6f,14f);p.quadraticTo(11f,21f,15f,13f);drawCircle(tint,1f,Offset(14f,8f)) }
                "whale" -> { p.moveTo(2f,12f);p.cubicTo(2f,1f,17f,5f,17f,14f);p.lineTo(22f,9f);p.lineTo(22f,17f);p.cubicTo(10f,25f,2f,21f,2f,12f);p.close();p.moveTo(8f,5f);p.lineTo(8f,1f);p.moveTo(8f,3f);p.lineTo(12f,1f);drawCircle(tint,1f,Offset(6f,12f)) }
                "paw" -> { drawOval(tint,Offset(6f,12f),androidx.compose.ui.geometry.Size(12f,9f),style=stroke);for(i in 0..3)drawCircle(tint,2f,Offset(4f+i*5f,if(i==0||i==3)9f else 5f),style=stroke) }
                "thumb" -> { p.moveTo(5f,10f);p.lineTo(10f,10f);p.lineTo(13f,2f);p.lineTo(16f,3f);p.lineTo(15f,10f);p.lineTo(22f,10f);p.lineTo(20f,21f);p.lineTo(5f,21f);p.close() }
                "party" -> { p.moveTo(2f,22f);p.lineTo(7f,7f);p.lineTo(17f,17f);p.close();p.moveTo(10f,4f);p.lineTo(13f,1f);p.moveTo(17f,9f);p.lineTo(22f,5f);p.moveTo(20f,14f);p.lineTo(24f,13f) }
                "comet" -> { drawCircle(tint,5f,Offset(7f,17f),style=stroke);p.moveTo(3f,12f);p.lineTo(21f,2f);p.lineTo(12f,21f);p.moveTo(11f,13f);p.lineTo(21f,3f) }
                "surprise" -> { drawCircle(tint,9f,Offset(12f,12f),style=stroke);drawCircle(tint,1f,Offset(8f,9f));drawCircle(tint,1f,Offset(16f,9f));drawCircle(tint,3f,Offset(12f,16f),style=stroke) }
                "laugh" -> { drawCircle(tint,9f,Offset(12f,12f),style=stroke);p.moveTo(6f,10f);p.lineTo(8f,8f);p.lineTo(10f,10f);p.moveTo(14f,10f);p.lineTo(16f,8f);p.lineTo(18f,10f);p.moveTo(6f,14f);p.lineTo(18f,14f);p.quadraticTo(12f,25f,6f,14f) }
                "melt" -> { p.moveTo(3f,15f);p.cubicTo(-1f,0f,24f,0f,21f,15f);p.cubicTo(28f,19f,19f,24f,13f,20f);p.cubicTo(5f,26f,-3f,20f,3f,15f);drawCircle(tint,1f,Offset(8f,9f));drawCircle(tint,1f,Offset(16f,11f));p.moveTo(8f,15f);p.quadraticTo(13f,20f,17f,14f) }
                else -> { drawCircle(tint,8f,Offset(12f,12f),style=stroke);drawCircle(tint,1f,Offset(9f,10f));drawCircle(tint,1f,Offset(15f,10f));p.moveTo(8f,15f);p.quadraticTo(12f,20f,16f,15f) }
            }
            drawPath(p,tint,style=stroke)
        }
    }
}
@Composable
internal fun Avatar(value: UByte,modifier: Modifier = Modifier) {
    val raw=value.toInt();val animal=animals.getOrElse(raw and 15){"paw"}
    val base=if(raw==0)Color(0xFFB4B2A9) else palette[raw shr 4]
    val tint=if(MaterialTheme.colorScheme.background.luminance()>.5f)Color(base.red*.45f,base.green*.45f,base.blue*.45f) else base
    Box(modifier.size(44.dp).background(tint.copy(alpha=.12f),CircleShape),contentAlignment=Alignment.Center){MeshIcon(animal,tint)}
}
